### ISSUE TITLE
About the use of mask on objective.py

### ISSUE THREAD

**Issue by wfcb-85** [google-research/simclr#20]
Hello,
I would like to know the logic behind the use of the mask on the 'add_contrastive_loss' function:
It is my understanding that you use it in order to decrease the value of the similarity for s_{i,i} when computing the loss.  
What you do is the following:

**LARGE_NUM = 1e9
masks = masks = tf.one_hot(tf.range(batch_size), batch_size)
logits_aa = tf.matmul(hidden1, hidden1_large, transpose_b=True) / temperature
 logits_aa = logits_aa - masks * LARGE_NUM_**

I know that the final value for the logits of (i,i) will be a very large negative number. But wouldnt it be more precise to just set the logits of positions (i,i) to zero ? I was thinking of the following code: 

**logits_aa = tf.matmul(hidden1, hidden1_large, transpose_b=True) / temperature
mask_inv = tf.one_hot(tf.range(batch_size), batch_size , on_value=0, off_value=1)
logits_aa = logits_aa * mask_inv**

Thanks in advance

**Comment by mimbres** [comment]
I just tried this, and it seems to work (in different tasks).
`
N = 5 # N: number of original samples in the batch

logits_aa = np.reshape(np.arange(N*N), (N, N)) # fake logits_aa...
masks = tf.constant(1 - tf.eye(N))
_logits_aa = tf.boolean_mask(logits_aa, masks)
_logits_aa = tf.reshape(_logits_aa, (N, N-1))

labels = tf.one_hot(tf.range(N), N * 2 - 1)

print('logits_aa:\n', logits_aa)
print('masks:\n', masks)
print('_logits_aa:\n', _logits_aa)
print('labels:\n', labels)
'

> logits_aa:
>  [[ 0  1  2  3  4]
>  [ 5  6  7  8  9]
>  [10 11 12 13 14]
>  [15 16 17 18 19]
>  [20 21 22 23 24]]
> masks:
>  tf.Tensor(
> [[0. 1. 1. 1. 1.]
>  [1. 0. 1. 1. 1.]
>  [1. 1. 0. 1. 1.]
>  [1. 1. 1. 0. 1.]
>  [1. 1. 1. 1. 0.]], shape=(5, 5), dtype=float32)
> _logits_aa:
>  tf.Tensor(
> [[ 1  2  3  4]
>  [ 5  7  8  9]
>  [10 11 13 14]
>  [15 16 17 19]
>  [20 21 22 23]], shape=(5, 4), dtype=int64)
> labels:
>  tf.Tensor(
> [[1. 0. 0. 0. 0. 0. 0.]
>  [0. 1. 0. 0. 0. 0. 0.]
>  [0. 0. 1. 0. 0. 0. 0.]
>  [0. 0. 0. 1. 0. 0. 0.]], shape=(4, 7), dtype=float32)

**Comment by chentingpc** [comment]
Masking logits with zero is different from masking with large negative. This is due to exp(0)=1 when you compute softmax with the logits, unlike exp(-LARGE_NUM)=0.
